decoder/stats.go

```diff
@@ -108,18 +104,82 @@ func ReloadGeofenceAndClearStats() {
 	pokemonCount = make(map[geo.AreaName]*areaPokemonCountDetail) // clear count
 }
 
-func updatePokemonStats(old *Pokemon, new *Pokemon, areaNames []geo.AreaName) {
-	var areas []geo.AreaName
+// update stats for an encounterId
+func updateEncounterStats(pokemon *Pokemon) {
+	// We should only be called from encounters. It's important to do so,
+	// so that the 'DuplicateEncounters' stats below are correct.
+	// And double check that we have IVs, anyway.
+	if !(pokemon.AtkIv.Valid && pokemon.DefIv.Valid && pokemon.StaIv.Valid) {
+		return
+	}
+
+	// Keep track of encounter Id -> account username. Count shinies
+	// for the same encounter Ids, but only if an account has not seen
+	// it before. We'll ignore things like re-rolls.
+
+	username := pokemon.Username.ValueOrZero()
+	if username == "" {
+		username = "<NoUsername>"
+	}
+
+	encounterCacheVal := encounterCache.GetOrCreate(pokemon.Id)
+	isNewEncounter := encounterCacheVal.NumAccountsSeen() == 0
+
+	if encounterCacheVal.SetAccountSeen(pokemon.Username.ValueOrZero()) {
```

Comment:
If I have understood correctly I think this is showing whether a single account makes a duplicate encounter.
I wonder if an interesting check of MITM optimal workflow would be if the encounter has been seen by any account before (the mystery scanner passes this test, lol)

Reply:
Yeah. I think it was you who had made a comment on discord that we gain the ability to have a metric on this... So, I added it. The metric for 'a different account has seen it before' is here too. I did them as labels... So you could just add both values for the label to get a total of all duplicate encounters.

Reply:
So, it appears this scanner does mostly pass the test. :) Duplicate encounters went away... EXCEPT... on the hour, every hour... there's a lot. What happens on the hour? Is some cache incorrectly purging things every hour?

![Screenshot 2023-11-30 at 7 07 54 PM](https://github.com/UnownHash/Golbat/assets/637328/6c09e61f-47dd-4e54-8247-e10c84a41011)

Reply:
That's more a question for the other project. :) But that graph is from the code in this PR.

Reply:
Weather changes

Reply:
Yeah, I wasn't thinking. I thought of that, but I'm going to blame lack of sleep on me thinking weather changes were every 15min. :) Thnx